contracts/src/mixins/ERC4337Mixin.sol

```diff
@@ -0,0 +1,244 @@
+
+        // --- DUPLICATE TRANSACTION SIGNATURE CHECK ---
+        // Revert if this signer has already submitted *any* signature for this *exact* safeTxHash
+        require(
+            !_signerSignedTx(safeTxHash, signer),
+            SignerAlreadySignedTransaction(signer, safeTxHash)
+        );
+
+        _verifySignature(safeTxHash, userOp.signature, signer, r, vs);
+
+        uint256 nonce = getNonce(signer);
+        require(userOp.nonce == nonce, UnexpectedNonce(nonce));
+
+        // We trust the entry point to set the correct prefund value, based on the operation params
+        // We need to perform this even if the signature is not valid, else the simulation function of the entry point will not work.
+        if (missingAccountFunds != 0) {
+            // We intentionally ignore errors in paying the missing account funds, as the entry point is responsible for
+            // verifying the prefund has been paid. This behaviour matches the reference base account implementation.
+            (bool success, ) = payable(msg.sender).call{
+                value: missingAccountFunds
+            }("");
+            (success);
+        }
+
+        return _packValidationData(false, 0, 0);
+    }
+
+    function _verifySafeTxHash(
+        bytes calldata callData
+    ) private pure returns (bytes32, address, bytes32, bytes32) {
+        (
+            bytes32 safeTxHash,
+            address safeAddress,
+            uint256 chainId,
+            uint256 nonce,
+            address to,
+            uint256 value,
+            bytes memory data,
+            uint8 operation,
+            uint256 safeTxGas,
+            uint256 baseGas,
+            uint256 gasPrice,
+            address gasToken,
+            address refundReceiver,
+            address signer,
+            bytes32 r,
+            bytes32 vs
+        ) = abi.decode(
+                callData,
+                (
+                    bytes32,
+                    address,
+                    uint256,
+                    uint256,
+                    address,
+                    uint256,
+                    bytes,
+                    uint8,
+                    uint256,
+                    uint256,
+                    uint256,
+                    address,
+                    address,
+                    address,
+                    bytes32,
+                    bytes32
+                )
+            );
+        bytes32 computedSafeTxHash = CoreLib._computeSafeTxHash(
+            safeAddress,
+            chainId,
+            nonce,
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver
+        );
+
+        require(
+            computedSafeTxHash == safeTxHash,
+            UnexpectedSafeTxHash(computedSafeTxHash)
+        );
+        return (safeTxHash, signer, r, vs);
```

Comment:
Not for this PR (as it is related to another optimisation comment), but it feels a bit weird that we use `r || s || v` signature format in some places, and `r || vs` in others - I would stick to just one.

Nothing to do in this PR itself.